python/mspasspy/db/schema.py

```diff
@@ -597,5 +597,83 @@ def set_writeable(self, key):
             raise MsPASSError(key + ' is not defined', 'Invalid')
         self._main_dic[key]['readonly'] = False
 
+
+'''
+check if a mspass.yaml file is valid or not
+'''
+def is_basic_type(s):
+    if s in ["ObjectID","int","float","double","complex","bool","string","list","dict","bytes"]:
+        return True
+    return False
+
+def get_all_db_name(schema_dic):
+    db_name_list = []
+    for db in schema_dic:
+        db_name_list.append(db)
+    return db_name_list
+
+def get_all_collection_name_by_db(schema_dic, db):
+    collection_name_list = []
+    if db not in schema_dic:
+        return collection_name_list
+    for c in schema_dic[db]:
+        collection_name_list.append(c)
+    return collection_name_list
+
 def _check_format(schema_dic):
-    pass
+    db_name_list = get_all_db_name(schema_dic)
+    collection_name_list = []
+    for db_name in db_name_list:
+        collection_name_list += get_all_collection_name_by_db(schema_dic, db_name)
+    
+    type_attribute_schema = schema.Schema({
+        'type':schema.And(str, lambda s: is_basic_type(s)),
+        schema.Optional('reference'):schema.And(str, lambda s: s in collection_name_list),
+        schema.Optional('concept'):str,
+        schema.Optional('aliases'):schema.Or(list, str),
+        schema.Optional('optional'):bool,
+        schema.Optional('readonly'):bool
+    }, ignore_extra_keys=True)
+    
+    reference_attribute_schema = schema.Schema({
+        'reference':schema.And(str, lambda s: s in collection_name_list),
+        schema.Optional('optional'):bool
+    }, ignore_extra_keys=True)
+    
+    metadata_attribute_schema = schema.Schema({
+        'collection':schema.And(str, lambda s: s in collection_name_list),
+        schema.Optional('readonly'):bool
+    }, ignore_extra_keys=True)
+    
+    database_collection_schema = schema.Schema({
+        schema.Optional('base'):schema.And(str, lambda s: s in collection_name_list),
+        'schema':schema.And(dict, schema.And(dict, lambda dic: ('_id' in dic and type_attribute_schema.validate(attr) 
+                                        if 'type' in dic[attr] else reference_attribute_schema.validate(attr) 
+                                        for attr in dic)))
+    }, ignore_extra_keys=True)
+    
+    metadata_collection_schema = schema.Schema({
+        'schema':schema.And(dict, lambda dic: (metadata_attribute_schema.validate(attribute) for attribute in dic))
+    }, ignore_extra_keys=True)
+    
+    database_schema = schema.Schema({
+        'wf_TimeSeries':schema.And(dict, lambda dic: database_collection_schema.validate(dic)),
+        'wf_Seismogram':schema.And(dict, lambda dic: database_collection_schema.validate(dic)),
```

Comment:
wf_TimeSeries and wf_Seismogram are actually not required, instead, what's required really is a collection that is the default `wf`. e.g.:
```
Database:
  wf_something:
    default: wf
```
The rule is that we need one collection that has `default: wf` defined, or if there is no `default: wf`, we will then need a collection that is named as `wf` itself (because `wf` itself will be the default name when not specified).

This is my bad - another implicit rule that I didn't mention anywhere. Actually, `elog` and `history_object` are also required default names, just like the `wf`.

Reply:
It's ok, I should have asked you about this first:) so from my understanding, a very basic schema should at least contain a collection called `wf` or a collection that has `default` attribute which equals to `wf`. Then `elog` and `history_object` are collections that related to `wf`.

Reply:
Exactly. The rule for `wf` also applies to `elog` and `history_object`.